operator/controllers/instance_controller.go

```diff
@@ -0,0 +1,461 @@
+		if errors.IsNotFound(err) {
+			instanceName := req.NamespacedName.Name
+			helmInstanceName := fmt.Sprintf("metabroker-%s", instanceName)
+			namespace := req.NamespacedName.Namespace
+			podName := fmt.Sprintf("metabroker-%s-deprovision", instanceName)
+			return r.deprovisioningPod(ctx, helmInstanceName, podName, namespace)
+		}
+		return ctrl.Result{}, err
+	}
+
+	planNamespacedName := req.NamespacedName
+	planNamespacedName.Name = instance.Spec.Plan
+	plan := &servicebrokerv1alpha1.Plan{}
+	if err := r.Get(ctx, planNamespacedName, plan); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	instanceNeedsUpdate := false
+
+	if len(instance.OwnerReferences) == 0 {
+		if err := r.setOwnership(ctx, instance, plan); err != nil {
+			return ctrl.Result{}, nil
+		}
+		instanceNeedsUpdate = true
+	}
+
+	if instance.Spec.ID == "" {
+		id := uuid.Must(uuid.NewUUID()) // UUID v1
+		instance.Spec.ID = id.String()
+		instanceNeedsUpdate = true
+	}
+
+	if instance.Spec.ValidateValues == nil {
+		validateValues := true
+		instance.Spec.ValidateValues = &validateValues
+		instanceNeedsUpdate = true
+	}
+
+	if instanceNeedsUpdate {
+		if err := r.Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// TODO: update the status with validated so the controller doesn't keep performing the
+	// validation on requeues.
+	if *instance.Spec.ValidateValues {
+		schemaJSON, err := yaml.YAMLToJSON([]byte(plan.Spec.Provisioning.Values.Schema))
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		schema := gojsonschema.NewBytesLoader(schemaJSON)
+		valuesJSON, err := yaml.YAMLToJSON([]byte(instance.Spec.Values))
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		values := gojsonschema.NewBytesLoader(valuesJSON)
+		result, err := gojsonschema.Validate(schema, values)
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		if !result.Valid() {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values. Including specific errors returned from
+			// result.Errors().
+			return ctrl.Result{}, nil
+		}
+	}
+
+	valuesSecretName := fmt.Sprintf("metabroker-%s-values", instance.Name)
+	if created, err := r.valuesSecret(ctx, instance, valuesSecretName, instance.Namespace); err != nil {
+		return ctrl.Result{}, err
+	} else if created {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	helmInstanceName := fmt.Sprintf("metabroker-%s", instance.Name)
+	podName := fmt.Sprintf("metabroker-%s-provision", instance.Name)
+	return r.provisioningPod(ctx, instance, plan, helmInstanceName, podName, instance.Namespace, valuesSecretName)
+}
+
+func (r *InstanceReconciler) setOwnership(
```

Comment:
I think in general this should be call `setController` (or if you're okay with the verbosity, `setControllingOwner`) rather than `setOwnership`, because this isn't just setting an owner reference. (This also applies to the other instance we already have.)

Reply:
True - that indeed makes more sense.

Reply:
I'll do the change for all the methods in a single follow-up PR.

Reply:
Actually, I fixed this one up here: f825f04. And fixed the others in https://github.com/SUSE/metabroker/pull/10 (which got merged before this PR).